sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -405,6 +409,246 @@ protected EventProcessor()
+                        || ex is ThreadAbortException)
+                    {
+                        throw;
+                    }
+                    catch (Exception ex)
+                    {
+                        ++failedConsumerCount;
+                        capturedException = ex;
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            if (consumer != null)
+                            {
+                                await consumer.CloseAsync(CancellationToken.None).ConfigureAwait(false);
+                            }
+                        }
+                        catch (Exception ex)
+                        {
+                            Logger.EventProcessorPartitionProcessingError(partition.PartitionId, Identifier, EventHubName, ConsumerGroup, ex.Message);
+
+                            // Do not bubble the exception, as the consumer is being refreshed; failure to close this consumer is non-fatal.
+                        }
+                    }
+                }
+
+                // If there was an exception captured, then surface it.  Otherwise signal that cancellation took place.
+
+                if (capturedException != null)
+                {
+                    ExceptionDispatchInfo.Capture(capturedException).Throw();
+                }
+
+                throw new TaskCanceledException();
+            }
+
+            // Start processing in the background and return the processor
+            // metadata.
+
+            return new PartitionProcessor
+            (
+                Task.Run(performProcessing, cancellationSource.Token),
+                readLastEnquedEventInformation,
+                cancellationSource
+            );
+        }
+
+        /// <summary>
+        ///   Performs the tasks needed to process a batch of events.
+        /// </summary>
+        ///
+        /// <param name="partition">The Event Hub partition whose processing should be started.</param>
+        /// <param name="eventBatch">The batch of events to process.</param>
+        /// <param name="dispatchEmptyBatches"><c>true</c> if empty batches should be dispatched to the handler; otherwise, <c>false</c>.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the processing.</param>
+        ///
+        internal virtual async Task ProcessEventBatchAsync(TPartition partition,
+                                                           IEnumerable<EventData> eventBatch,
+                                                           bool dispatchEmptyBatches,
+                                                           CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            eventBatch = (eventBatch as IList<EventData>) ?? eventBatch.ToList();
```

Comment:
This should come back as `IList` unless we introduce a different protocol/consumer in the future.   We want to be sure that we can safely enumerate the set multiple times for instrumentation and `Any` checks.

Reply:
> Had you considered changing the return value of TransportConsumer.ReceiveAsync to be Task<IReadOnlyList<EventData>> instead of Task<IEnumerable<EventData>>?

So... yeah.  That's the right thing to do in hindsight, seeing as how we control the set being built after AMQP translation.  I'll admit to being afraid that it would cause a good amount of churn, but the surface area was smaller than expected.   Going to go in this direction.

Reply:
@kinelski -  This will impact you for the partition receiver stuff, unless you merge before I close this one out.